### PR TITLE
Fix encoding on locales with non-arabic digits (as Hindi (India))

### PR DESCRIPTION
**Pull request by EonTechnology**
Unit tests for Bencode on different locales (if needed):
```java
package com.dampcake.bencode;

import static org.junit.Assert.assertEquals;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.ConcurrentSkipListMap;

import org.junit.AfterClass;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

/**
 * Unit tests for Bencode on different locales.
 */
@SuppressWarnings("unchecked")
@RunWith(Parameterized.class)
public class BencodeLocaleTest {

    private static Locale startLocale = Locale.getDefault();

    @Rule
    public ExpectedException exception = ExpectedException.none();

    @Parameterized.Parameter
    public Locale testLocale;

    private Bencode instance;

    @AfterClass
    public static void restore() throws Exception {
        Locale.setDefault(startLocale);
    }

    @Parameterized.Parameters(name = "{index}: {0}")
    public static Collection<Object[]> data() {
        LinkedList<Object[]> res = new LinkedList<Object[]>();
        for (Locale locale : Locale.getAvailableLocales()) {
            res.add(new Object[] {locale});
        }
        return res;
    }

    @Before
    public void setUp() {
        Locale.setDefault(testLocale);
        instance = new Bencode();
    }

    @Test
    public void testDecodeDictionary() throws Exception {
        Map<String, Object> decoded = instance.decode(
                "d4:dictd3:1234:test3:4565:thinge4:listl11:list-item-111:list-item-2e6:numberi123456e6:string5:valuee".getBytes(),
                Type.DICTIONARY);

        assertEquals(4, decoded.size());

        assertEquals("value", decoded.get("string"));
        assertEquals(123456L, decoded.get("number"));

        List<Object> list = (List<Object>) decoded.get("list");
        assertEquals(2, list.size());
        assertEquals("list-item-1", list.get(0));
        assertEquals("list-item-2", list.get(1));

        Map<String, Object> map = (Map<String, Object>) decoded.get("dict");
        assertEquals(2, map.size());
        assertEquals("test", map.get("123"));
        assertEquals("thing", map.get("456"));
    }

    @Test
    public void testWriteDictionary() throws Exception {
        byte[] encoded = instance.encode(new HashMap<Object, Object>() {{
            put("string", "value");
            put("number", 123456);
            put("list", new ArrayList<Object>() {{
                add("list-item-1");
                add("list-item-2");
            }});
            put("dict", new ConcurrentSkipListMap() {{
                put(123, "test");
                put(456, "thing");
            }});
        }});

        assertEquals(
                "d4:dictd3:1234:test3:4565:thinge4:listl11:list-item-111:list-item-2e6:numberi123456e6:string5:valuee",
                new String(encoded, instance.getCharset()));
    }

    @Test
    public void testDecodeEncodedDictionary() throws Exception {
        byte[] encoded = instance.encode(new HashMap<Object, Object>() {{
            put("string", "value");
            put("number", 123456);
            put("list", new ArrayList<Object>() {{
                add("list-item-1");
                add("list-item-2");
            }});
            put("dict", new ConcurrentSkipListMap() {{
                put(123, "test");
                put(456, "thing");
            }});
        }});

        Map<String, Object> decoded = instance.decode(encoded, Type.DICTIONARY);

        assertEquals(4, decoded.size());

        assertEquals("value", decoded.get("string"));
        assertEquals(123456L, decoded.get("number"));

        List<Object> list = (List<Object>) decoded.get("list");
        assertEquals(2, list.size());
        assertEquals("list-item-1", list.get(0));
        assertEquals("list-item-2", list.get(1));

        Map<String, Object> map = (Map<String, Object>) decoded.get("dict");
        assertEquals(2, map.size());
        assertEquals("test", map.get("123"));
        assertEquals("thing", map.get("456"));
    }
}
```

### CODE REVIEW

File: src/main/java/com/dampcake/bencode/BencodeOutputStream.java [hunk 1]
```diff
@@ -126,13 +126,13 @@ public void writeDictionary(final Map<?, ?> m) throws IOException {
     private static String encode(final String s) {
         if (s == null) throw new NullPointerException("s cannot be null");
 
-        return String.format("%d%s%s", s.length(), Bencode.SEPARATOR, s);
+        return String.format("%s%s%s", Integer.toString(s.length()), Bencode.SEPARATOR, s);
```
Comment (dampcake) [inline src/main/java/com/dampcake/bencode/BencodeOutputStream.java]: What does this do? When would `s.legnth()` ever return anything that would need to be toString'd?

File: src/main/java/com/dampcake/bencode/BencodeOutputStream.java [hunk 2]
```diff
@@ -126,13 +126,13 @@ public void writeDictionary(final Map<?, ?> m) throws IOException {
...
     }
 
     private static String encode(final Number n) {
         if (n == null) throw new NullPointerException("n cannot be null");
 
-        return String.format("%s%d%s", Bencode.NUMBER, n.longValue(), Bencode.TERMINATOR);
+        return String.format("%s%s%s", Bencode.NUMBER, Long.toString(n.longValue()), Bencode.TERMINATOR);
```
Comment (dampcake) [inline src/main/java/com/dampcake/bencode/BencodeOutputStream.java]: Bencode states that numbers should be stored in base10. When would n.longValue (a long) need to be converted to a string? This seems like something that should be done on the client side.

File: src/main/java/com/dampcake/bencode/BencodeOutputStream.java (hunk 1 above)
Comment (EonTechnology) [reply src/main/java/com/dampcake/bencode/BencodeOutputStream.java]: `String.format("%d", ...)` uses locale digits.
`Integer.toString(...)` uses "normal" digits.
Or locale may be set on method calls: `String.format(Locale.ENGLISH, "%d", ...)`.

Comment (dampcake) [reply src/main/java/com/dampcake/bencode/BencodeOutputStream.java]: Makes sense. I think I would rather use the Locale.ENGLISH solution as that looks to be the more accepted solution. Also if you could update the BencodeOutputStreamTest to run parameterized with all the locales that would be awesome!

**Comment by dampcake** [comment]
I was hoping to get this released to maven today. Looks like it will take a little while longer (troubleshooting with sonatype right now).

For now you can use the SNAPSHOT if you need it right away.

**Pull request merged**